src/main/java/io/ebean/datasource/pool/ConnectionPool.java

```diff
@@ -392,49 +379,53 @@ protected void notifyWarning(String msg) {
     }
   }
 
-  private synchronized void notifyDataSourceIsDown(SQLException ex) {
-
-    if (dataSourceUp) {
+  private void notifyDataSourceIsDown(SQLException reason) {
+    if (dataSourceUp.get()) {
       reset();
-    }
-    dataSourceUp = false;
-    if (ex != null) {
-      dataSourceDownReason = ex;
-    }
-    if (!dataSourceDownAlertSent) {
-      dataSourceDownAlertSent = true;
-      logger.error("FATAL: DataSourcePool [" + name + "] is down or has network error!!!", ex);
-      if (notify != null) {
-        notify.dataSourceDown(this, ex);
-      }
+      notifyDown(reason);
     }
   }
 
-  private synchronized void notifyDataSourceIsUp() {
-    if (dataSourceDownAlertSent) {
-      // set to false here, so that a getConnection() call in DataSourceAlert.dataSourceUp
-      // in same thread does not fire the event again (and end in recursion)
-      // all other threads will be blocked, becasue method is synchronized.
-      dataSourceDownAlertSent = false;
-      logger.error("RESOLVED FATAL: DataSourcePool [" + name + "] is back up!");
-      if (notify != null) {
-        notify.dataSourceUp(this);
+  private void notifyDown(SQLException reason) {
+    synchronized (dataSourceUp) {
+      if (dataSourceUp.get()) {
+        // check and set false immediately so that we only alert once
+        dataSourceUp.set(false);
```

Comment:
@rbygrave  can we use `if (dataSourceUp.getAndSet(false))` here and avoid synchronization completely?